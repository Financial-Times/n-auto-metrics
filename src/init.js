let metrics;

export const initAutoMetrics = instance => {
	metrics = instance;
};

export const getMetricsInstance = () => metrics;
