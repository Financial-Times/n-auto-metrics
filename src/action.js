import { getMetricsInstance } from './init';
import { isPromise } from './utils';

const createEventMetrics = ({ namespace, operation, action }) => {
	const metrics = getMetricsInstance();
	if (metrics === undefined) {
		throw Error(`auto metrics instance needs to be initialised first`);
	}
	const serviceRoot = `service.${namespace}.${action}`;
	const operationRoot = `operation.${operation}.action.${action}`;
	const countOne = path => {
		metrics.count(`${serviceRoot}.${path}`, 1);
		metrics.count(`${operationRoot}.${path}`, 1);
	};
	return {
		start: () => countOne('start'),
		success: () => countOne('success'),
		failure: e => {
			countOne(`failure.status.${e.status}`);
			countOne(`failure.type.${e.type}`);
		},
	};
};

const metricsEvent = ({ namespace, operation, action }) => {
	const event = createEventMetrics({ namespace, operation, action });
	event.start();
	return event;
};

export const autoMetricsAction = callFunction => (
	paramsOrArgs = {},
	meta = {},
) => {
	const namespace = meta.namespace || paramsOrArgs.namespace || 'action';
	const action = callFunction.name;
	if (typeof namespace !== 'string' || namespace.includes(' ')) {
		throw Error(
			`action metrics namespace needs to be string without spaces, at function ${action}`,
		);
	}
	const operation = meta.operation || paramsOrArgs.operation;

	const event = metricsEvent({ namespace, operation, action });

	try {
		const call = callFunction(paramsOrArgs, meta);
		if (isPromise(call)) {
			return call
				.then(data => {
					event.success();
					return data;
				})
				.catch(e => {
					event.failure(e);
					throw e;
				});
		}
		const data = call;
		event.success();
		return data;
	} catch (e) {
		event.failure(e);
		throw e;
	}
};

export const autoMetricsActions = serviceName => actionBundle => {
	if (typeof serviceName !== 'string' || serviceName.includes(' ')) {
		throw Error(`service name for metrics needs to be string without spaces`);
	}
	const enhanced = {};
	Object.keys(actionBundle).forEach(methodName => {
		enhanced[methodName] = (paramsOrArgs, meta) =>
			autoMetricsAction(actionBundle[methodName])(paramsOrArgs, {
				...meta,
				namespace: serviceName,
			});
	});
	return enhanced;
};
