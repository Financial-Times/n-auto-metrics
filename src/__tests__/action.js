import { initAutoMetrics } from '../init';
import { autoMetricsAction, autoMetricsActions } from '../action';

const metrics = {
	count: jest.fn(),
};

describe('autoMetricsAction', () => {
	beforeAll(() => {
		initAutoMetrics(metrics);
	});

	afterEach(() => {
		jest.resetAllMocks();
	});

	describe('should throw error', () => {
		it('if metrics was not initialised', async () => {
			initAutoMetrics(undefined);
			const callFunction = () => null;
			const execution = autoMetricsAction(callFunction);
			expect(execution).toThrowErrorMatchingSnapshot();
			initAutoMetrics(metrics);
		});

		it('namespace is not string', async () => {
			const callFunction = () => null;
			const execution = () => autoMetricsAction(callFunction)({ namespace: 1 });
			expect(execution).toThrowErrorMatchingSnapshot();
		});

		it('namespace has space', async () => {
			const callFunction = () => null;
			const execution = () =>
				autoMetricsAction(callFunction)({ namespace: 'foo bar' });
			expect(execution).toThrowErrorMatchingSnapshot();
		});
	});

	describe('enhance async function', () => {
		it('should invoke callFunction correctly', async () => {
			const callFunction = jest.fn(() => Promise.resolve('foo'));
			const params = { test: 'a' };
			const meta = { meta: 'b' };
			const result = await autoMetricsAction(callFunction)(params, meta);
			expect(callFunction.mock.calls).toMatchSnapshot();
			const expectedResult = await callFunction(params, meta);
			expect(result).toBe(expectedResult);
		});

		it('should log callFunction success correctly', async () => {
			const callFunction = () => Promise.resolve('foo');
			await autoMetricsAction(callFunction)();
			expect(metrics.count.mock.calls).toMatchSnapshot();
		});

		it('should log callFunction failure correctly and throw the original exception', async () => {
			const errorInstance = { message: 'bar' };
			const callFunction = async () => {
				throw errorInstance;
			};
			try {
				await autoMetricsAction(callFunction)();
			} catch (e) {
				expect(e).toBe(errorInstance);
				expect(metrics.count.mock.calls).toMatchSnapshot();
			}
		});
	});

	describe('non-async function', () => {
		it('should invoke callFunction correctly', () => {
			const callFunction = jest.fn(() => 'foo');
			const params = { test: 'a' };
			const meta = { meta: 'b' };
			const result = autoMetricsAction(callFunction)(params, meta);
			expect(callFunction.mock.calls).toMatchSnapshot();
			const expectedResult = callFunction(params, meta);
			expect(result).toBe(expectedResult);
		});

		it('should log callFunction success correctly', () => {
			const callFunction = () => 'foo';
			autoMetricsAction(callFunction)();
			expect(metrics.count.mock.calls).toMatchSnapshot();
		});

		it('should log callFunction failure correctly and throw the original exception', () => {
			const errorInstance = { message: 'bar' };
			const callFunction = () => {
				throw errorInstance;
			};
			try {
				autoMetricsAction(callFunction)();
			} catch (e) {
				expect(e).toBe(errorInstance);
				expect(metrics.count.mock.calls).toMatchSnapshot();
			}
		});
	});

	describe('args format (paramsOrArgs, meta)', () => {
		it('takes the namespace/operation value in meta first if available', () => {
			const callFunction = () => null;
			const params = { a: 'test' };
			const meta = { namespace: 'foo', operation: 'bar' };
			autoMetricsAction(callFunction)(params, meta);
			expect(metrics.count.mock.calls).toMatchSnapshot();
		});

		it('takes the namespace/operation value in paramsOrArgs if not available in meta', () => {
			const callFunction = () => null;
			const args = { a: 'test', namespace: 'foo', operation: 'bar' };
			autoMetricsAction(callFunction)(args);
			expect(metrics.count.mock.calls).toMatchSnapshot();
		});

		it('uses default namespace "action" and undefined operation if not available in args', () => {
			const callFunction = () => null;
			const args = { a: 'test' };
			autoMetricsAction(callFunction)(args);
			expect(metrics.count.mock.calls).toMatchSnapshot();
		});
	});
});

describe('autoMetricsActions', () => {
	afterEach(() => {
		jest.resetAllMocks();
	});

	it('decorate each method correctly', async () => {
		const callFunctionA = jest.fn();
		const callFunctionB = jest.fn();
		const enhancedService = autoMetricsActions('mock-service')({
			callFunctionA,
			callFunctionB,
		});
		const paramsA = { test: 'a' };
		const paramsB = { test: 'b' };
		const meta = { operation: 'mock-operation' };
		await enhancedService.callFunctionA(paramsA, meta);
		await enhancedService.callFunctionB(paramsB, meta);
		expect(callFunctionA.mock.calls).toMatchSnapshot();
		expect(callFunctionB.mock.calls).toMatchSnapshot();
		expect(metrics.count.mock.calls).toMatchSnapshot();
	});
});
