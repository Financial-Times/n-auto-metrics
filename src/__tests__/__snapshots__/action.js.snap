// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoMetricsAction args format (paramsOrArgs, meta) takes the namespace/operation value in meta first if available 1`] = `
Array [
  Array [
    "service.foo.callFunction.start",
    1,
  ],
  Array [
    "operation.bar.action.callFunction.start",
    1,
  ],
  Array [
    "service.foo.callFunction.success",
    1,
  ],
  Array [
    "operation.bar.action.callFunction.success",
    1,
  ],
]
`;

exports[`autoMetricsAction args format (paramsOrArgs, meta) takes the namespace/operation value in paramsOrArgs if not available in meta 1`] = `
Array [
  Array [
    "service.foo.callFunction.start",
    1,
  ],
  Array [
    "operation.bar.action.callFunction.start",
    1,
  ],
  Array [
    "service.foo.callFunction.success",
    1,
  ],
  Array [
    "operation.bar.action.callFunction.success",
    1,
  ],
]
`;

exports[`autoMetricsAction args format (paramsOrArgs, meta) uses default namespace "action" and undefined operation if not available in args 1`] = `
Array [
  Array [
    "service.action.callFunction.start",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.start",
    1,
  ],
  Array [
    "service.action.callFunction.success",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.success",
    1,
  ],
]
`;

exports[`autoMetricsAction enhance async function should invoke callFunction correctly 1`] = `
Array [
  Array [
    Object {
      "test": "a",
    },
    Object {
      "meta": "b",
    },
  ],
]
`;

exports[`autoMetricsAction enhance async function should log callFunction failure correctly and throw the original exception 1`] = `
Array [
  Array [
    "service.action.callFunction.start",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.start",
    1,
  ],
  Array [
    "service.action.callFunction.failure.status.undefined",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.failure.status.undefined",
    1,
  ],
  Array [
    "service.action.callFunction.failure.type.undefined",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.failure.type.undefined",
    1,
  ],
]
`;

exports[`autoMetricsAction enhance async function should log callFunction success correctly 1`] = `
Array [
  Array [
    "service.action.callFunction.start",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.start",
    1,
  ],
  Array [
    "service.action.callFunction.success",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.success",
    1,
  ],
]
`;

exports[`autoMetricsAction non-async function should invoke callFunction correctly 1`] = `
Array [
  Array [
    Object {
      "test": "a",
    },
    Object {
      "meta": "b",
    },
  ],
]
`;

exports[`autoMetricsAction non-async function should log callFunction failure correctly and throw the original exception 1`] = `
Array [
  Array [
    "service.action.callFunction.start",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.start",
    1,
  ],
  Array [
    "service.action.callFunction.failure.status.undefined",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.failure.status.undefined",
    1,
  ],
  Array [
    "service.action.callFunction.failure.type.undefined",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.failure.type.undefined",
    1,
  ],
]
`;

exports[`autoMetricsAction non-async function should log callFunction success correctly 1`] = `
Array [
  Array [
    "service.action.callFunction.start",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.start",
    1,
  ],
  Array [
    "service.action.callFunction.success",
    1,
  ],
  Array [
    "operation.undefined.action.callFunction.success",
    1,
  ],
]
`;

exports[`autoMetricsAction should throw error if metrics was not initialised 1`] = `"auto metrics instance needs to be initialised first"`;

exports[`autoMetricsAction should throw error namespace has space 1`] = `"action metrics namespace needs to be string without spaces, at function callFunction"`;

exports[`autoMetricsAction should throw error namespace is not string 1`] = `"action metrics namespace needs to be string without spaces, at function callFunction"`;

exports[`autoMetricsActions decorate each method correctly 1`] = `
Array [
  Array [
    Object {
      "test": "a",
    },
    Object {
      "namespace": "mock-service",
      "operation": "mock-operation",
    },
  ],
]
`;

exports[`autoMetricsActions decorate each method correctly 2`] = `
Array [
  Array [
    Object {
      "test": "b",
    },
    Object {
      "namespace": "mock-service",
      "operation": "mock-operation",
    },
  ],
]
`;

exports[`autoMetricsActions decorate each method correctly 3`] = `
Array [
  Array [
    "service.mock-service.mockConstructor.start",
    1,
  ],
  Array [
    "operation.mock-operation.action.mockConstructor.start",
    1,
  ],
  Array [
    "service.mock-service.mockConstructor.success",
    1,
  ],
  Array [
    "operation.mock-operation.action.mockConstructor.success",
    1,
  ],
  Array [
    "service.mock-service.mockConstructor.start",
    1,
  ],
  Array [
    "operation.mock-operation.action.mockConstructor.start",
    1,
  ],
  Array [
    "service.mock-service.mockConstructor.success",
    1,
  ],
  Array [
    "operation.mock-operation.action.mockConstructor.success",
    1,
  ],
]
`;
